### PR TITLE
#127 Making player data more... sort of strongly typed, kind of

### PR DESCRIPTION
**Pull request by EdgarVerona**
To note, where the change is feels a little weird in there because the MainMenu scene *has to* be initialized before the join game is called, otherwise the per-device-type handling from there down is never initialized and you end up with a blank screen on game join.

I tested it out and it works, but just to note in case it's odd how it initializes the relevant MainMenu scene and then immediately attempts to see if it should call join_game.

### CODE REVIEW

File: common/lobby/Lobby.gd [hunk 1]
```diff
@@ -34,13 +34,13 @@ func populate_map_list():
...
-	var existingPlayer := find_player_node(playerId)
-	if existingPlayer != null:
+	var existingPlayerNode := find_player_node(playerId)
+	if existingPlayerNode != null:
 		return
 	
 	print("Creating player in lobby")
 	
-	var player = GameData.players[playerId]
+	var player := GameData.players[playerId] as PlayerData
```
Comment (Wavesonics) [inline common/lobby/Lobby.gd]: This could use: `GameData.get_player(playerId)`

File: common/lobby/Lobby.gd [hunk 2]
```diff
@@ -76,7 +76,7 @@ func remove_player(playerId: int):
 
 
 func repopulate_player(playerId: int):
-	var player = GameData.players[playerId]
+	var player := GameData.players[playerId] as PlayerData
```
Comment (Wavesonics) [inline common/lobby/Lobby.gd]: Use `GameData.get_player(playerId)`

File: networking/GameData.gd [hunk 3]
```diff
@@ -23,25 +18,27 @@ func get_players() -> Array:
...
+		is_host = false
 	}
 
-
-func add_player(newPlayer: Dictionary) -> bool:
-	var playerId = newPlayer[PLAYER_ID]
+func add_player_from_raw_data(newPlayerDictionary: Dictionary) -> bool:
+	var playerId = newPlayerDictionary.id
 	if not self.players.has(playerId):
+		var newPlayer := load("res://networking/PlayerData.gd").new() as PlayerData
```
Comment (Wavesonics) [inline networking/GameData.gd]: Change `load()` to `preload()`

File: networking/GameData.gd [hunk 4]
```diff
@@ -52,32 +49,33 @@ func reset():
...
 
-func update_player(player):
-	var playerId = player[PLAYER_ID]
-	players[playerId] = player
-
+func update_player_from_raw_data(player_data_dictionary: Dictionary):
+	var playerId = player_data_dictionary.id
+	get_player(playerId).load(player_data_dictionary)
 
-func get_host() -> Player:
```
Comment (Wavesonics) [inline networking/GameData.gd:76]: lawl bugs

File: server/lobby/ServerLobby.gd [hunk 5]
```diff
@@ -67,7 +67,7 @@ func get_name() -> String:
 # Parse command line port in the form of:
 # --port xxxxx
 func get_port() -> int:
-	var port := ServerNetwork.SERVER_PORT
```
Comment (Wavesonics) [inline server/lobby/ServerLobby.gd]: why can't this be strongly typed?

**Review by Wavesonics — CHANGES_REQUESTED** [review]

File: networking/GameData.gd (hunk 4 above)
Comment (EdgarVerona) [reply networking/GameData.gd:76]: Ha, but it worked!  Thanks to strange typing!

File: server/lobby/ServerLobby.gd (hunk 5 above)
Comment (EdgarVerona) [reply server/lobby/ServerLobby.gd]: I don't know actually, it was a seemingly unrelated error message I got while debugging.  I am actually not sure if it's already a bug even without my changes - I imagine it must be.  But I was seeing this error come up when i was resolving my other errors.

**Review by Wavesonics — APPROVED** [review]

**Pull request merged**